app/controllers/orders_controller.rb

```diff
@@ -0,0 +1,90 @@
+class OrdersController < ApplicationController
+  before_action :require_login, only: [:index]
+  before_action :correct_user, only: [:index]
+  before_action :order_items, only: [:new, :create]
+  before_action :find_user, except: [:new, :create, :confirm, :cancel]
+
+  def index
+    orders = @user.orders
+    @orders = orders.uniq.sort
+    @paid_orders = Order.filter_orders(@orders, "paid")
+    @completed_orders = Order.filter_orders(@orders, "shipped")
+    @cancelled_orders = Order.filter_orders(@orders, "cancelled")
+    @paid_orders_revenue = Order.user_orders_revenue(@paid_orders, @user)
+    @completed_orders_revenue = Order.user_orders_revenue(@completed_orders, @user)
+    @total_revenue = @paid_orders_revenue + @completed_orders_revenue
+    if !(params["Status"].nil? || params["Status"] == "")
+      @orders = Order.filter_orders(@orders, params["Status"])
+    end
+  end
+
+  def show
+    @order = Order.find(params[:id])
+  end
+
+  def new
+    not_enough_stock = false
+    @order_items.each do |oi|
+      not_enough_stock = true if Product.find(oi.product_id).stock < oi.quantity
+    end
+    if not_enough_stock
+      flash[:error] = "There is not enough stock to complete your purchase. Please update your cart."
+      redirect_to cart_path
+    else
+      @order = Order.new
+    end
+  end
+
+  def create
+    @order = Order.new(order_params)
+    products_sold = []
+    @order_items.each do |oi|
+      @order.order_items << oi
+      products_sold.push([oi.product, oi.quantity])
+      oi.save
+    end
+    if @order.save
+      Product.decrement_stock(products_sold)
```

Comment:
There's a very small chance that even though we've saved our order successfully, the process of decrementing our stock inventories according will actually fail. In that scenario, we will have sold some items without updating our inventory counts, and our site would then allow later users to over-order those items.

We can avoid this by using ActiveRecord's [transaction method](http://vaidehijoshi.github.io/blog/2015/08/18/safer-sql-using-activerecord-transactions/). How might this controller action's code look differently if we used transactions?